base/compiler/tfuncs.jl

```diff
@@ -724,20 +724,34 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
     elseif isconstType(headtypetype)
         headtype = headtypetype.parameters[1]
     else
-        return Any
+        return Type
     end
     largs = length(args)
     if headtype === Union
         largs == 0 && return Const(Bottom)
-        largs == 1 && return args[1]
+        hasnonType = false
         for i = 1:largs
             ai = args[i]
-            if !isa(ai, Const) || !isa(ai.val, Type)
+            if isa(ai, Const)
+                if !isa(ai.val, Type)
+                    if isa(ai.val, TypeVar) || isa(ai.val, PartialTypeVar)
+                        return hasnonType = true
```

Comment:
Did you mean like so, @vtjnash?

Reply:
Bump. I'm still not sure you meant this. A `Const` `PartialTypeVar` looks a bit funny here to me.

Reply:
`isa(ai.val, PartialTypeVar)` can return `Union{}`. But `ai` itself can be a `PartialTypeVar`

Reply:
Alright, then I've apparently misinterpreted your earlier remark. The `isa(ai, PartialTypeVar)` case is already handled below (in the `!isType(ai)` and `!isa(ai, Type)` branch).